# Hands-on Exercise for  FPM Module

### 1. Exploring properties of the dataset accidents_10k.dat. Read more about it here:  http://fimi.uantwerpen.be/data/accidents.pdf

In [1]:
!head accidents_10k.dat

1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 
2 5 7 8 9 10 12 13 14 15 16 17 18 20 22 23 24 25 27 28 29 32 33 34 35 36 37 38 39 
7 10 12 13 14 15 16 17 18 20 25 28 29 30 33 40 41 42 43 44 45 46 47 48 49 50 51 52 
1 5 8 10 12 14 15 16 17 18 19 20 21 22 24 25 26 27 28 29 30 31 41 43 46 48 49 51 52 53 54 55 56 57 58 59 60 61 
5 8 10 12 14 15 16 17 18 21 22 24 25 26 27 28 29 31 33 36 38 39 41 43 46 56 62 63 64 65 66 67 68 
7 8 10 12 17 18 21 23 24 26 27 28 29 30 33 34 35 36 38 41 43 47 59 63 66 69 70 71 72 73 74 75 76 77 78 79 
1 12 14 15 16 17 18 21 22 23 24 25 27 28 29 30 31 35 38 41 43 44 53 56 57 58 59 60 63 66 80 81 82 83 84 
10 12 14 15 16 17 18 21 22 24 25 26 27 28 29 30 31 33 39 41 43 44 46 49 59 60 62 63 66 82 
1 8 10 12 14 15 16 17 18 21 22 23 24 25 27 29 30 31 38 41 43 53 56 59 61 63 66 68 85 86 87 88 89 
1 8 12 13 14 15 16 17 18 22 24 25 28 30 38 41 42 43 46 49 60 63 64 66 80 82 84 90 91 92 93 94 95 


<span style="color:red">**Question 1a:** </span>. How many items are there in the data?

In [2]:
!awk -- '{for (i = 1; i <= NF; i++) wc[$i] += 1}; END {print length(wc)}' accidents_10k.dat 

310


<span style="color:green">**Answer:** </span>itemsets :310

<span style="color:red">**Question 1b:** </span> How many transactions are present in the data?

In [4]:
!wc -L accidents_10k.dat

145 accidents_10k.dat


<span style="color:green">**Answer:** </span> transactions : 10000

<span style="color:red">**Question 1c:** </span>.  What is the length of the smallest transaction?

In [19]:
!awk '(NR==1||length< smallest){smallest=length} END {print smallest}' accidents_10k.dat

69


<span style="color:green">**Answer:** </span> length of smallest transaction : 69

<span style="color:red">**Question 1d:** </span>  What is the length of the longest transaction?

In [1]:
!awk '(NR==1||length> longest){longest=length} END {print longest}' accidents_10k.dat

145


<span style="color:green">**Answer:** </span>  length of smallest transaction : 145

<span style="color:red">**Question 1e:** </span>  What is the size of the search space of frequent itemsets in this data?

In [39]:
!chmod u+x apriori

<span style="color:green">**Answer:** </span> search space :2^itemsets = 2^310

<span style="color:red">**Question 1f:** </span> 
Assume that you work for the deparment of transportation that collected this data. What benefit do you see in using itemset mining approaches on this data?

<span style="color:green">**Answer:** </span> we can analyze the sequence of accidents that occured using the data. prevention measure are taken where they are large accidents and etc.

<span style="color:red">**Question 1g:** </span>  What type of itemsets (frequent, maximial or closed) would you be interested in discovering this dataset? State your reason.

<span style="color:green">**Answer:** </span> finding the closed itemsets is interested. as we can get frequent itemsets and also its support.

<span style="color:red">**Question 1h:** </span>  What minsup threshold would you use and why?

In [ ]:
the minsup threshold i would use is 1500 . it depends upon the support and confidence of the itemset.

### 2. Generating frequent, maximal and closed itemsets using $\color{red}{\text{Apriori}}$, $\color{red}{\text{ECLAT}}$, and $\color{red}{\text{FPGrowth}}$ algorihtms from the dataset accidents_10k.dat 

<span style="color:red">**Question 2a:** </span> Generate frequent itemsets using Apriori, for minsup = 2000, 3000, and 4000. Which of these minsup thresholds results in a maximum number of frequent itemsets? Which of these minsup thresholds results in a least number of frequent itemsets? Provide a rationale for these observations.

In [41]:
!chmod u+x apriori

In [42]:
!./apriori

usage: ./apriori [options] infile [outfile]
find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
-t#      target type                              (default: s)
         (s: frequent, c: closed, m: maximal item sets,
          g: generators, r: association rules)
-m#      minimum number of items per set/rule     (default: 1)
-n#      maximum number of items per set/rule     (default: no limit)
-s#      minimum support of an item set/rule      (default: 10%)
-S#      maximum support of an item set/rule      (default: 100%)
         (positive: percentage, negative: absolute number)
-o       use original rule support definition     (body & head)
-c#      minimum confidence of an assoc. rule     (default: 80%)
-e#      additional evaluation measure            (default: none)
-a#      aggregation mode for evaluation measure  (default: none)
-d#      threshold for add. evaluation measure    (default: 10%)
-i       invalidate eva

In [43]:
!./apriori -ts -s-2000 accidents_10k.dat T_Accidents_2000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.01s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [19.75s].
writing T_Accidents_2000.txt ... [851034 set(s)] done [0.10s].


In [44]:
!./apriori -ts -s-4000 accidents_10k.dat T_Accidents_4000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [33 item(s)] done [0.01s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.00s].
building transaction tree ... [22267 node(s)] done [0.01s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [1.39s].
writing T_Accidents_4000.txt ... [29501 set(s)] done [0.00s].


In [45]:
!./apriori -ts -s-3000 accidents_10k.dat T_Accidents_3000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [38 item(s)] done [0.00s].
sorting and reducing transactions ... [9674/10000 transaction(s)] done [0.01s].
building transaction tree ... [24741 node(s)] done [0.01s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 done [4.38s].
writing T_Accidents_3000.txt ... [133799 set(s)] done [0.01s].


<span style="color:green">**Answer:** </span> the maximum number of frequent set resulted is 851034 when minsup threshold is 2000 

the minimum number of frequent sets resulted is 29501 for minsup = 4000 

as the minsup threshold increases the resultant number of frequent set decreases

<span style="color:red">**Question 2b:** </span>   Using Apriori, compare the execution time for finding frequent itemsets for minsup = 2000, 3000, and 4000. Which of these minsup thresholds takes the least amount of time? Provide a rationale for this observation.

In [47]:
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-2000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.00s].
building transaction tree ... [20250 node(s)] done [0.01s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [18.30s].
writing <null> ... [851034 set(s)] done [0.02s].
18 secs  611808 microsecs


In [48]:
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-3000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.03s].
filtering, sorting and recoding items ... [38 item(s)] done [0.00s].
sorting and reducing transactions ... [9674/10000 transaction(s)] done [0.01s].
building transaction tree ... [24741 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 done [4.50s].
writing <null> ... [133799 set(s)] done [0.00s].
5 secs  421554 microsecs


In [49]:
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-4000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.01s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.00s].
building transaction tree ... [22267 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [1.33s].
writing <null> ... [29501 set(s)] done [0.00s].
1 secs  675855 microsecs


<span style="color:green">**Answer:** </span> For minsup = 4000 the execution takes the least amount of time and minsup = 1000 takes more amount of time. 

observation : from the above results the minsup threshold increases and the execution time decreases. since there are more frequent item sets when the minsup is low it takes more time when the threshold is less. its inversely proportional

<span style="color:red">**Question 2c:** </span> Using Apriori, find the frequent itemsets for minsup = 2000, 3000, and 4000. Determine the number of itemsets for each size (1 to max length of an itemset). What trends do you see that are common for all three minsup thresholds? What trends do you see that are different? Provide a rationale for these observations.

In [23]:
print("*****************minsup = 2000******************************")
!awk '{print NF-1}' T_Accidents_2000.txt|sort -n|uniq -c
print("*****************minsup = 3000******************************")
!awk '{print NF-1}' T_Accidents_3000.txt|sort -n|uniq -c
print("*****************minsup = 4000******************************")
!awk '{print NF-1}' T_Accidents_4000.txt|sort -n|uniq -c

*****************minsup = 2000******************************
     49 1
    705 2
   5285 3
  23745 4
  69647 5
 139628 6
 195730 7
 193299 8
 133819 9
  63937 10
  20497 11
   4189 12
    483 13
     21 14
*****************minsup = 3000******************************
     38 1
    468 2
   2830 3
   9887 4
  21779 5
  31964 6
  32020 7
  21862 8
   9839 9
   2705 10
    387 11
     20 12
*****************minsup = 4000******************************
     33 1
    319 2
   1492 3
   4043 4
   6926 5
   7751 6
   5626 7
   2546 8
    668 9
     91 10
      6 11


<span style="color:green">**Answer:** </span> 
the number of itemsets decreases as minsup threshold increases from the above result. and the number of frequent sets also decreases for each item as minsup threshold increases.

for each minup threshold the number of itemsets follows the bell curve

<span style="color:red">**Question 2d:** </span>  Using Apriori with minsup=2000, compare the number of frequent, maximal, and closed itemsets. Which is the largest set and which is the smallest set? Provide a rationale for these observations.

In [30]:
print("*****************frequent item set******************************")
!./apriori -ts -s-2000 accidents_10k.dat T_Accidents_frequent_2000.txt
print("*****************closed item set******************************")
!./apriori -tc -s-2000 accidents_10k.dat T_Accidents_closed_2000.txt
print("*****************maximal item set******************************")
!./apriori -tm -s-2000 T10I4D100K_new.dat T_Accidents_maximal_2000.txt

*****************frequent item set******************************
./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [19.37s].
writing T_Accidents_frequent_2000.txt ... [851034 set(s)] done [0.10s].
*****************closed item set******************************
./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.03s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transaction

<span style="color:green">**Answer:** </span> the frequent item set is the largest and the maximal itemset is smallest.

the maximal item set (subset of) closed item set (subset of) frequent itemset 

<span style="color:red">**Question 2e:** </span> For a minsup = 2000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [ ]:
!chmod u+x eclat
!chmod u+x fpgrowth

In [65]:
import datetime 
algo = {"apriori","eclat","fpgrowth"}
for i in algo:
    start = datetime.datetime.now()
    ! printf "*****************""$i""*************************\n"
    !./"$i" -ts -s-2000 accidents_10k.dat 
    end = datetime.datetime.now()
    elapsed = end - start
    print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

*****************apriori*************************
./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.01s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.00s].
building transaction tree ... [20250 node(s)] done [0.01s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [19.88s].
writing <null> ... [851034 set(s)] done [0.01s].
20 secs  395722 microsecs
*****************fpgrowth*************************
./fpgrowth - find frequent item sets with the fpgrowth algorithm
version 6.18 (2019.03.31)        (c) 2004-2019   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 tran

<span style="color:green">**Answer:** </span> the FP growth algorithm tooks the least amount of time for the execution. as the computing time of fp growth algorithm is less than other algorithms.


<span style="color:red">**Question 2f:** </span> For a minsup = 4000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [66]:
import datetime 
algo = {"apriori","eclat","fpgrowth"}
for i in algo:
    ! printf "*****************""$i""*************************\n"
    start = datetime.datetime.now()
    !./"$i" -ts -s-4000 accidents_10k.dat 
    end = datetime.datetime.now()
    elapsed = end - start
    print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

*****************apriori*************************
./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.01s].
building transaction tree ... [22267 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [1.16s].
writing <null> ... [29501 set(s)] done [0.00s].
1 secs  473746 microsecs
*****************fpgrowth*************************
./fpgrowth - find frequent item sets with the fpgrowth algorithm
version 6.18 (2019.03.31)        (c) 2004-2019   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s

<span style="color:green">**Answer:** </span> when the minsup threshold = 4000 the execution time is least for fpgrowth algorithm. it is beacuase of the computing time is efficient for the algorithm. the time difference between fp growth and eclat is less


<span style="color:red">**Question 2g:** </span>  For a minsup = 6000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [63]:
import datetime 
algo = {"apriori","eclat","fpgrowth"}
for i in algo:
    ! printf "*****************""$i""*************************\n"
    start = datetime.datetime.now()
    !./"$i" -ts -s-6000 accidents_10k.dat 
    end = datetime.datetime.now()
    elapsed = end - start
    print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

*****************apriori*************************
./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [20 item(s)] done [0.00s].
sorting and reducing transactions ... [3216/10000 transaction(s)] done [0.01s].
building transaction tree ... [6478 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 done [0.04s].
writing <null> ... [2254 set(s)] done [0.00s].
0 secs  306305 microsecs
*****************fpgrowth*************************
./fpgrowth - find frequent item sets with the fpgrowth algorithm
version 6.18 (2019.03.31)        (c) 2004-2019   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [20 item(s)] done [0.00s].
sorting and reducing transactions ... [3216/10000 transaction(s)] done [0

<span style="color:green">**Answer:** </span>  when minsup = 6000 the  fpgrowth algorithm works faster but its alsmost equal to eclat algorithm as the difference is of nanosecs.

moreover as the minsup increases the execution time decreases and 3 algorithms are working efficient.

<span style="color:red">**Question 2h:** </span> Fill the following table based on execution times computed in __2e__, __2f__, and __2g__. State your observations on the relative computational efficiency at different support thresholds. Based on your knowledge of these algorithms, provide the reasons behind your observations.

In [67]:
import datetime 
algo = {"apriori","eclat","fpgrowth"}
minsup ={2000,4000,6000}
for i in algo:
    ! printf "*****************""$i""*************************\n"
    for j in minsup:
        ! printf "*****************""$j""*************************\n"
        start = datetime.datetime.now()
        !./"$i" -ts -s-"$j" accidents_10k.dat 
        end = datetime.datetime.now()
        elapsed = end - start
        print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

*****************apriori*************************
*****************6000*************************
./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [20 item(s)] done [0.00s].
sorting and reducing transactions ... [3216/10000 transaction(s)] done [0.00s].
building transaction tree ... [6478 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 done [0.04s].
writing <null> ... [2254 set(s)] done [0.00s].
0 secs  319735 microsecs
*****************4000*************************
./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transaction

<span style="color:green">Answer: </span>
    
|   Algorithm                |minsup=2000                   |minsup=4000                      |minsup=6000                 |
|----------------------------|------------------------------|---------------------------------|----------------------------| 
|Apriori                     | 19 secs  402198 microsecs    |     1 secs  288746 microsecs    |   0 secs  319735 microsecs |
|Eclat                       | 0 secs  532076 microsecs     |     0 secs  342196 microsecs    |   0 secs  283213 microsecs |
|FPGrowth                    | 0 secs  418950 microsecs     |     0 secs  299857 microsecs    |   0 secs  282791 microsecs |


<span style="color:green">**Answer:** </span> from the above comparisions the FP growth algorithm  works faster for the given data set when minsup = 2000,4000 and 6000 ,also as minsup threshold increase it works even more faster as it contains less number of itemsets.

 As the minup increases the difference in time between the algorithm decreases which you can clearly observe at minsup = 6000
 


### 3. Discovering frequent subsequences and substrings

Assume that roads in a Cincinnati are assigned numbers. Participants are enrolled in a transportation study and for every trip they make using their car, the sequence of roads taken are recorded. Trips that involves freeways are excluded. This data is in the file <span style="color:blue">road_seq_data.dat</span>.

In [37]:
!chmod u+x prefixspan

<span style="color:red">**Question 3a:** </span>  What 'type' of sequence mining will you perform to determine frequently taken 'paths'? Paths are sequences of roads traveresed consecutively in the same order.

<span style="color:green">**Answer:** </span> We perform sequence mining here as the transportation of students is different and anaysis using the subsequence patterns will do good for the analysis.

<span style="color:red">**Question 3b:** </span> How many sequences are there in this sequence database?

In [7]:
!wc -l road_seq_data.dat

1000 road_seq_data.dat


<span style="color:green">**Answer:** </span> 1000 sequences

<span style="color:red">**Question 3c:** </span> What is the size of the alphabet in this sequence database?

In [6]:
!awk -- '{for (alphabet = 1; alphabet <= NF; alphabet++) wc[$alphabet] += 1}; END {print length(wc)}' road_seq_data.dat

1283


<span style="color:green">**Answer:** </span>  the size of alphabet is the number of items in the sequence database - 1283

<span style="color:red">**Question 3d:** </span> What are the total number of possible subsequences of length 2 in this dataset?

<span style="color:red">**Question 3e:** </span> What are the total number of possible substrings of length 2 in this dataset?

In [8]:
!chmod u+x seqwog

there will be 1282 substring possible as there are 1283 items are present in the dataset

<span style="color:red">**Question 3f:** </span> Discover frequent __subsequences__ with minsup = 10 and report the number of subsequences discovered.

In [38]:
!./prefixspan -ts -s-10 road_seq_data.dat road_seq_data_minsup_10_prefix.txt

./prefixspan - sequence mining with unique occurrences of items and weight averaging
version 2.16 (2016.10.15)        (c) 2010-2016   Christian Borgelt
reading road_seq_data.dat ... [1283 item(s), 1000 transaction(s)] done [0.01s].
recoding items ... [1283 item(s)] done [0.00s].
filtering and reducing transactions ... [910/1000 transaction(s)] done [0.00s].
writing road_seq_data_minsup_10_prefix.txt ... [4588 sequence(s)] done [0.04s].


<span style="color:green">**Answer:** </span> there are 4588 frequent subsequnces when minsup = 3 

<span style="color:red">**Question 3g:** </span>  Discover frequent __substrings__ with minsup = 10 and report the number of substrings discovered.

In [16]:
!./seqwog -ts -s-10 road_seq_data.dat road_seq_data_minsup_10.txt

./seqwog - find frequent sequences without gaps
version 3.16 (2016.10.15)        (c) 2010-2016   Christian Borgelt
reading road_seq_data.dat ... [1283 item(s), 1000 transaction(s)] done [0.00s].
recoding items ... [1283 item(s)] done [0.00s].
reducing and triming transactions ... [844/1000 transaction(s)] done [0.00s].
writing road_seq_data_minsup_10.txt ... [613 sequence(s)] done [0.01s].


<span style="color:green">**Answer:** </span> 613 substrings are discovered

<span style="color:red">**Question 3h:** </span> Explain the difference in the number of frequent subsequences and substrings found in __3f__ and __3g__ above.

<span style="color:green">**Answer:** </span> the frequent subsequences are more than the frequent substrings as the substring are the immediate items but the subsequences can be anywhere but follows the order in the given set. so subsequences are generally more than the substrings in the dataset